db/write_thread.h

```diff
@@ -49,32 +49,74 @@ class WriteThread {
     // the leader to STATE_COMPLETED.
     STATE_GROUP_LEADER = 2,
 
-    // A Writer that has returned as a follower in a parallel group.
-    // It should apply its batch to the memtable and then call
-    // CompleteParallelWorker.  When someone calls ExitAsBatchGroupLeader
-    // or EarlyExitParallelGroup this state will get transitioned to
-    // STATE_COMPLETED.
-    STATE_PARALLEL_FOLLOWER = 4,
+    // The state used to inform a waiting writer that it has become the
+    // leader of memtable writer group. The leader will either write
+    // memtable for the whole group, or launch a parallel group write
+    // to memtable by calling LaunchParallelMemTableWrite.
+    STATE_MEMTABLE_WRITER_LEADER = 4,
+
+    // The state used to inform a waiting writer that it has become a
+    // parallel memtable writer. It can be the group leader who launch the
+    // praallel writer group, or one of the followers. The writer should then
+    // apply its batch to the memtable concurrently and call
+    // CompleteParallelMemTableWriter.
+    STATE_PARALLEL_MEMTABLE_WRITER = 8,
 
     // A follower whose writes have been applied, or a parallel leader
     // whose followers have all finished their work.  This is a terminal
     // state.
-    STATE_COMPLETED = 8,
+    STATE_COMPLETED = 16,
 
     // A state indicating that the thread may be waiting using StateMutex()
     // and StateCondVar()
-    STATE_LOCKED_WAITING = 16,
+    STATE_LOCKED_WAITING = 32,
   };
 
   struct Writer;
 
-  struct ParallelGroup {
-    Writer* leader;
-    Writer* last_writer;
+  struct WriteGroup {
+    Writer* leader = nullptr;
+    Writer* last_writer = nullptr;
     SequenceNumber last_sequence;
     // before running goes to zero, status needs leader->StateMutex()
     Status status;
     std::atomic<uint32_t> running;
+    size_t size = 0;
+
+    struct Iterator {
+      Writer* writer;
+      Writer* last_writer;
+
+      explicit Iterator(Writer* w, Writer* last)
+          : writer(w), last_writer(last) {}
+
+      Writer* operator*() const { return writer; }
+
+      Iterator& operator++() {
+        assert(writer != nullptr);
+        if (writer == last_writer) {
+          writer = nullptr;
+        } else {
+          writer = writer->link_newer;
+        }
+        return *this;
+      }
+
+      bool operator!=(const Iterator& other) const {
+        return writer != other.writer;
+      }
+    };
+
+    Iterator begin() const { return Iterator(leader, last_writer); }
+    Iterator end() const { return Iterator(nullptr, nullptr); }
+
+    autovector<Writer*> ToVector() {
+      autovector<Writer*> v;
+      for (auto w : *this) {
+        v.push_back(w);
+      }
+      return v;
```

Comment:
My C++ knowledge is poor. Will C++11 compilers guarantee that the copy of return value is emitted? I read this page and it is still not clear to me: http://en.cppreference.com/w/cpp/language/return#Notes

Do you know the sure answer of it? If not, I suggest we use the typical way of this code base, which is 
```
void ToVector(autovector<Writer*>* v)
```